draft-wang-ppm-differential-privacy.md

```diff
@@ -414,23 +459,204 @@ random seed into a sample from a given distribution.
+However, in case the minimum batch size is not met when Aggregators want to
+output their aggregate shares, each Aggregator can sample the same randomized
+response Client-DP on the "missing" Clients with zero-hot vectors, so the
+central DP guarantee can still be satisfied, based on the privacy amplification
+analysis in {{FMT20}} and {{FMT22}}.
+
+~~~
+class Prio3SumVecWithRandomizedResponse:
+    Measurement = List[int]
+    AggregateShare = List[Field]
+    AggregateResult = List[int]
+    DebiasedAggregateResult = List[float]
+
+    def __init__(self,
+                 rr_primitive: DpPrimitive,
+                 num_aggregators: Unsigned):
+        self.rr_primitive = rr_primitive
+        self.num_aggregators = num_aggregators
+
+    def maybe_add_noise_to_measurement(self,
+                                       meas: Measurement,
+                                       ) -> Measurement:
+        """
+        Apply randomized response primitive to measurement.
+        """
+        return self.rr_primitive.add_noise(meas)
+
+    def maybe_add_noise_to_agg_share(self,
+                                     agg_share: AggregateShare,
+                                     meas_count: Unsigned,
+                                     min_batch_size: Unsigned,
+                                     ) -> AggregateShare:
+        """
+        If minimum batch size is met, this function does nothing,
+        because DP guarantee would have already been met.
+        Otherwise, add the same randomized response noise on the
+        missing Clients.
+        """
+        if meas_count >= batch_size:
+            return agg_share
+        noise = self.rr_primitive.sample_noise(
+            batch_size - meas_count, len(agg_share)
+        )
+        return [a + Field(b) for (a, b) in zip(agg_share, noise)]
+
+    def maybe_debias(self,
+                     agg_result: AggregateResult,
+                     meas_count: Unsigned,
+                     min_batch_size: Unsigned,
+                     ) -> DebiasedAggregateResult:
+        """
+        Debias aggregate result with randomized response.
+        """
+        if meas_count >= min_batch_size:
+            return self.rr_primitive.debias(agg_result, meas_count)
+        # The count passed to debiasing will be:
+        # the number of actual Clients + the "missing" Clients
+        # sampled by all Aggregators.
+        debiasing_count = \
+            meas_count + \
+            num_aggregators * (min_batch_size - meas_count)
+        return self.rr_primitive.debias(agg_result, debiasing_count)
+~~~
+
+### Prio3Histogram with Discrete Gaussian
+
+For Prio3Histogram, we will apply an Aggregator-only DP mechanism,
+that is implemented with discrete Gaussian.
+
+~~~
+class Prio3HistogramWithDiscreteGaussian:
+    Measurement = int
+    AggregateShare = List[Field]
+    AggregateResult = List[int]
+    DebiasedAggregateResult = List[int]
```

Comment:
DebiasedAggregateResult = AggregateResult
to make it more clean

Reply:
Done.